src/AMDGPUOffload.cpp

```diff
@@ -0,0 +1,679 @@
+
+    uint16_t get_machine() override { return EM_AMDGPU; }
+    uint32_t get_flags() override { return flags; }
+    uint32_t get_version() override { return EV_CURRENT; }
+    void append_dynamic(Section &dynamic) override {
+    }
+
+    uint64_t get_got_entry(Section &got, const Symbol &sym) override {
+        // Check if we already made a got entry for this symbol.
+        for (const Relocation &r : got.relocations()) {
+            if (r.get_symbol() == &sym && r.get_type() == R_AMDGPU_RELATIVE64) {
+                internal_assert(r.get_addend() == 0);
+                return r.get_offset();
+            }
+        }
+
+        uint64_t got_offset = got.contents_size();
+        got.append_contents((uint64_t)0);
+        got.add_relocation(Elf::Relocation(R_AMDGPU_RELATIVE64, got_offset, 0, &sym));
+        return got_offset;
+    }
+
+    bool needs_plt_entry(const Relocation &r) override {
+        return maybe_branch_inst(r.get_type());
+    }
+
+    Symbol add_plt_entry(const Symbol &sym, Section &plt, Section &got, const Symbol &got_sym) override {
+        if (got.contents_empty()) {
+            // The PLT hasn't been started, initialize it now.
+            plt.set_alignment(16);
+
+            std::vector<char> padding(64, (char)0);
+            // TODO: Make a .plt0 entry that supports lazy binding.
+            plt.set_contents(padding.begin(), padding.end());
+        }
+
+        static const uint8_t hexagon_plt1[] = {
+            0x00, 0x40, 0x00, 0x00, // { immext (#0) (Relocation:R_HEX_B32_PCREL_X)
+            0x0e, 0xc0, 0x49, 0x6a, //   r14 = add (pc, ##GOTn@PCREL) }  (Relocation:R_HEX_6_PCREL_X)
+            0x1c, 0xc0, 0x8e, 0x91, //   r28 = memw (r14)
+            0x00, 0xc0, 0x9c, 0x52, //   jumpr r28
+        };
+
+        debug(2) << "Adding PLT entry for symbol " << sym.get_name() << "\n";
+
+        // Add a GOT entry for this symbol.
+        uint64_t got_offset = got.contents_size();
+        got.append_contents((uint32_t)0);
+        got.add_relocation(Elf::Relocation(R_HEX_JMP_SLOT, got_offset, 0, &sym));
+
+        // Add the PLT code.
+        uint32_t plt_offset = plt.get_size();
+        plt.append_contents(hexagon_plt1, hexagon_plt1 + sizeof(hexagon_plt1));
+
+        plt.add_relocation(Relocation(R_HEX_B32_PCREL_X, plt_offset + 0, got_offset, &got_sym));
+        plt.add_relocation(Relocation(R_HEX_6_PCREL_X, plt_offset + 4, got_offset + 4, &got_sym));
+
+        // Make a symbol for the PLT entry.
+        Symbol plt_sym("plt_" + sym.get_name());
+        plt_sym
+            .set_type(Symbol::STT_FUNC)
+            .set_binding(Symbol::STB_LOCAL)
+            .define(&plt, plt_offset, sizeof(hexagon_plt1));
+
+        return plt_sym;
+    }
+
+    Relocation relocate(uint64_t fixup_offset, char *fixup_addr, uint64_t type,
+                        const Elf::Symbol *sym, uint64_t sym_offset, int64_t addend,
+                        Elf::Section &got) override {
+        if (type == R_HEX_32) {
```

Comment:
This needs correcting. I could not find where relocate() is used so could not see what its interface is to know what the right think here is.

How are relocations for absolute symbols handled? If the symbol is defined then they can be relocated and a R_AMDGPU_RELATIVE relocation generated. If they are to an external symbol then an absolute relocation needs to be generated. But this code seems to unconditionally assume the symbol is defined and does not relocate it. So I suspect I am misunderstanding the code.

I was thinking they would still need to be relocated, but would need a relocation symbol

Reply:
@dsharletg can you help?

Reply:
I think do_relocation should be done first to update the relocation otherwise absolute relocations for defined symbols will not work. I suspect the Hexagon version is broken. Has a .data section symbol initialized to the address of another .data section symbol been tested?

Then:

`if (type == R_AMDGPU_ABS64 && sym->is_defined()) {`
`  // Use this relocation for defined 64 bit absolute relocations.`
`  return Relocation(R_AMDGPU_RELATIVE64, fixup_offset, sym_offset + addend, nullptr);`
`} else if (type == R_AMDGPU_ABS32_LO || type == R_AMDGPU_ABS32_HI || type == R_AMDGPU_ABS32 || type == R_AMDGPU_ABS64) {`
`  // Need to add dynamic relocation for the symbol.`
`  return Relocation(type, fixup_offset, addend, sym);'
`}`

Reply:
I *think* we have global initializers that refer to other globals, e.g.: https://github.com/halide/Halide/blob/master/src/runtime/fake_thread_pool.cpp#L26

Here's where the code to handle defined symbols for relocations begins: https://github.com/halide/Halide/blob/master/src/Elf.cpp#L838

But yeah, it does seem like maybe there's a missing code path here...?

Keep in mind that our linker only needs to handle a subset of possible kinds of things that are arbitrary elf module might do, especially if the runtime for your target is simple (as I imagine it is for a GPU).

Reply:
@t-tye what to do if both the `if` conditionals are false?

Reply:
Need to add a final return Relocation();

Could simply this to simply:

```
if (type == R_AMDGPU_ABS64 && sym->is_defined())
  return Relocation(R_AMDGPU_RELATIVE64, fixup_offset, sym_offset + addend, nullptr);
if (type == R_AMDGPU_ABS32_LO || type == R_AMDGPU_ABS32_HI || type == R_AMDGPU_ABS32 || type == R_AMDGPU_ABS64)
  return Relocation(type, fixup_offset, addend, sym);
return Relocation();
```